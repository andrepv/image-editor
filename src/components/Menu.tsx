import React from "react";
import { ReactComponent as Crop } from "../assets/crop.svg";
import { ReactComponent as Flip } from "../assets/flip.svg";
import { ReactComponent as Draw } from "../assets/pencil.svg";
import { ReactComponent as Text } from "../assets/text.svg";
import { ReactComponent as Search } from "../assets/search.svg";

import { ReactComponent as Adjust } from "../assets/adjust.svg";
import Tooltip from "./Tooltip";
import useStore from "../hooks/useStore";
import { useObserver } from "mobx-react";
import { ModeName } from "../stores/canvasStore";

interface IMenuItems {
  icon: React.ReactElement;
  name: ModeName;
  handler: () => void;
  tooltip?: string;
}

const Menu: React.FC = () => {
  const { appStore, canvasStore, imageStore } = useStore();

  const handleClick = (modeName: ModeName) => {
    if (!imageStore.url && modeName !== "search") {
      return;
    }

    appStore.toggleToolbar(modeName);

    if (canvasStore.mode && canvasStore.scale !== 1) {
      if (canvasStore.mode !== "search") {
        canvasStore.setScale(1);
      }
    }

    if (!canvasStore.mode) {
      canvasStore.resetToBaseScale();
    }
  };

  const items: IMenuItems[] = [
    {
      icon: <Search />,
      name: "search",
      handler: () => handleClick("search"),
      tooltip: "Upload an image from Unsplash",
    },
    {
      icon: <Crop />,
      name: "crop",
      handler: () => handleClick("crop"),
    },
    {
      icon: <Flip />,
      name: "rotate",
      handler: () => handleClick("rotate"),
    },
    {
      icon: <Draw />,
      name: "drawing",
      handler: () => handleClick("drawing"),
    },
    {
      icon: <Text />,
      name: "text",
      handler: () => handleClick("text"),
    },
    {
      icon: <Adjust />,
      name: "effects",
      handler: () => handleClick("effects"),
    },
  ];
  return useObserver(() => (
    <section className="menu">
      {items.map((item, index) => {
        const tooltip = item.tooltip || item.name;
        return (
          <Tooltip key={index} content={tooltip} placement="right">
            <div
              className={`menu__item ${
                canvasStore.mode === item.name ? "menu__item_active" : ""
              }`}
              onClick={item.handler}
            >
              {item.icon}
            </div>
          </Tooltip>
        );
      })}
    </section>
  ));
};

export default Menu;