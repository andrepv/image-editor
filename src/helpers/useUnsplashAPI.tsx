import { useEffect, useState, useReducer } from "react";
import Unsplash, { toJson } from "unsplash-js";
import { accessKey } from "../appConstants";

export type Image = {
  thumbUrl: string;
  regularUrl: string;
  width: number;
  height: number;
}

type State = {
  isLoading: boolean;
  isError: boolean;
  images: Image[];
  totalPages: number;
}

type Action =
 | {type: "pending"}
 | {type: "success", payload: {images: Image[]; totalPages: number}}
 | {type: "error"};

const unsplash = new Unsplash({accessKey});

const initialState = {
  isLoading: false,
  isError: false,
  images: [],
  totalPages: 0,
};

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case "pending":
      return {
        ...state,
        isLoading: true,
      };
    case "success":
      return {
        isLoading: false,
        isError: false,
        images: action.payload.images,
        totalPages: action.payload.totalPages,
      };
    case "error":
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    default:
      throw new Error();
  }
}

export const useUnsplashAPI = ():
  [
    State & {currentPage: number},
    {
      loadImages: (inputValue: string) => void;
      goToNextPage: () => void;
      goToPrevPage: () => void;
    }
  ] => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [keyword, setKeyword] = useState("beach");
  const [currentPage, goToPage] = useState(1);
  const togglePage = (pageNum: number) => {
    setTimeout(() => {
      goToPage(pageNum);
    }, 200);
  };
  const goToNextPage = () => togglePage(currentPage + 1);
  const goToPrevPage = () => togglePage(currentPage - 1);
  const loadImages = (inputValue: string) => {
    const value = inputValue.trim();
    if (value.length > 2) {
      setKeyword(inputValue);
      goToPage(1);
    }
  };

  useEffect(() => {
    const fetchData = async () => {
      try {
        dispatch({type: "pending"});

        const response = await unsplash.search.photos(
          keyword, currentPage, 10, {orientation: "portrait"},
        );
        const data = await toJson(response);
        const images: Image[] = data.results.map((item: any) => {
          return {
            thumbUrl: item.urls.thumb,
            regularUrl: item.urls.regular,
            width: item.width,
            height: item.height,
          };
        });
        dispatch({
          type: "success",
          payload: {
            images,
            totalPages: data.total_pages,
          },
        });
      } catch (e) {
        dispatch({type: "error"});
      }
    };
    if (keyword.length > 2) {
      fetchData();
    }
  }, [keyword, currentPage]);
  return [{
    ...state,
    currentPage,
  }, {
    loadImages,
    goToNextPage,
    goToPrevPage,
  }];
};