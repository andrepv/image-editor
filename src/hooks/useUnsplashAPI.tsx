import { useEffect } from "react";
import Unsplash, { toJson } from "unsplash-js";
import { accessKey } from "../appConstants";
import useStore from "./useStore";
import { autorun } from "mobx";

export type Image = {
  url: string;
  width: number;
  height: number;
}

const unsplash = new Unsplash({accessKey});

const useUnsplashAPI = () => {
  const {searchStore} = useStore();

  const updatePage = (pageNum: number) => {
    setTimeout(() => {
      searchStore.goToPage(pageNum);
    }, 200);
  };
  const goToNextPage = () => updatePage(searchStore.currentPage + 1);
  const goToPrevPage = () => updatePage(searchStore.currentPage - 1);
  const loadImages = (inputValue: string) => {
    const value = inputValue.trim();
    if (value.length > 2) {
      searchStore.setKeyword(inputValue);
      searchStore.goToPage(1);
    }
  };

  useEffect(() => {
    if (searchStore.isLoaded) {
      return;
    }
    autorun(() => {
      const {currentPage, keyword} = searchStore;
      const fetchData = async () => {
        try {
          searchStore.showLoader();

          const response = await unsplash.search.photos(
            keyword, currentPage, 10,
          );
          const data = await toJson(response);
          const images: Image[] = data.results.map((item: any) => {
            return {
              url: item.urls.regular,
              width: item.width,
              height: item.height,
            };
          });
          searchStore.loadImages(images, data.total_pages);
        } catch (e) {
          searchStore.showError();
        }
      };
      if (keyword.length > 2) {
        fetchData();
      }
    });
  }, []);

  return {
    loadImages,
    goToNextPage,
    goToPrevPage,
  };
};

export default useUnsplashAPI;