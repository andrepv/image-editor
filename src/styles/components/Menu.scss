@import "../variables.scss";

.menu {
  grid-row-start: 1;
  grid-row-end: 3;
  background-color: $menu-bg;
  z-index: 2;
  display: flex;
  align-items: center;
  &__wrapper {
    width: 100%;
  }
  &__item {
    padding: 1.875vw;
    cursor: pointer;
    svg {
      width: 2.34375vw;
      height: 2.34375vw;
    }
    &:hover {
      background: $toolbar-bg;
      transition-duration: .3s;
    }
    &_active {
      background-color: $toolbar-bg;
    }
  }
}