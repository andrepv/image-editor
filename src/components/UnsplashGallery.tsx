import React, { useEffect, useRef, useState } from "react";
import Gallery, { PhotoProps, PhotoClickHandler } from "react-photo-gallery";

import { useUnsplashAPI, Image } from "../helpers/useUnsplashAPI";
import { ReactComponent as Search } from "../assets/search2.svg";
import { ReactComponent as Loader } from "../assets/loader.svg";

type Props = {
  close: () => void;
}

const UnsplashGallery: React.FC<Props> = ({close}) => {
  const [{
    images,
    isLoading,
    isError,
    currentPage,
    totalPages,
  }, {
    loadImages,
    goToNextPage,
    goToPrevPage,
  }] = useUnsplashAPI();
  const [inputValue, setInputValue] = useState("");
  const galleryRef = useRef<HTMLDivElement>(null);

  const handleClickOutside = (event: MouseEvent) => {
    if (!galleryRef || !galleryRef.current) {
      return;
    }
    const target = event.target as HTMLElement;
    if (!galleryRef.current.contains(target)) {
      close();
    }
  };

  const handleImageClick: PhotoClickHandler = (event, photos) => {
    // temporarily
    console.log(images[photos.index].regularUrl);
  };

  const galleryPhotoProp: PhotoProps[] = images.map((image: Image) => {
    return {
      width: image.width,
      height: image.height,
      src: image.thumbUrl,
    };
  });

  useEffect(() => {
    document.addEventListener("click", handleClickOutside);
    return () => {
      document.removeEventListener("click", handleClickOutside);
    };
  });

  return (
    <div
      ref={galleryRef}
      className="unsplash-gallery__container"
    >
      <div className="unsplash-gallery custom-scrollbar">
        <form
          className="unsplash-gallery__form"
          onSubmit={
            event => {
              loadImages(inputValue);
              event.preventDefault();
            }
          }
        >
          <button
            type="submit"
            className="unsplash-gallery__submit-btn"
          >
            <Search />
          </button>
          <input
            className="unsplash-gallery__input"
            type="text"
            value={inputValue}
            onChange={event => setInputValue(event.target.value)}
            placeholder="What are you looking for?"
          />
        </form>
        {images.length && !isError ? (
        <div className="unsplash-gallery__pagination">
          {currentPage !== 1 && (
            <button
              className="unsplash-gallery__prev-btn"
              onClick={goToPrevPage}
            >
              Previous
            </button>
          )}
          {currentPage !== totalPages && (
            <button
              className="unsplash-gallery__next-btn"
              onClick={goToNextPage}
            >
              Next
            </button>
          )}
        </div>
        ) : null}
        {isError && (
          <div className="unsplash-gallery__error">
            Something went wrong ...
          </div>
        )}
        {isLoading
        ? <div className="unsplash-gallery__loader"><Loader/></div>
        : images.length
          ? <Gallery photos={galleryPhotoProp} onClick={handleImageClick}/>
          : <div className="unsplash-gallery__error">No results</div>
        }
      </div>
    </div>
  );
};

export default UnsplashGallery;