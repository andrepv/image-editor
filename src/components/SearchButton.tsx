import React, { useState } from "react";

import { ReactComponent as Search } from "../assets/search.svg";
import UnsplashGallery from "./UnsplashGallery";
import Tooltip from "./Tooltip";

const SearchButton: React.FC = () => {
  const [isGalleryOpen, toggleGallery] = useState(false);
  const closeGallery = () => toggleGallery(false);
  return (
    <>
      <Tooltip content="Upload an image from Unsplash" placement="bottom">
        <Search
          className="header__search"
          onClick={() => toggleGallery(!isGalleryOpen)}
        />
      </Tooltip>
      {isGalleryOpen && <UnsplashGallery close={closeGallery} />}
    </>
  );
};

export default SearchButton;