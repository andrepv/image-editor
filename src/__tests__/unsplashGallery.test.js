import React from "react";
import ReactDOM from "react-dom";
import { shallow, mount } from "enzyme";

import SearchButton from "../components/SearchButton";
import UnsplashGallery from "../components/UnsplashGallery";
import Menu from "../components/Menu";
import useUnsplashAPI from "../helpers/useUnsplashAPI";

jest.mock("../helpers/useUnsplashAPI");

const data = {
  images: [],
  isLoading: false,
  isError: false,
  currentPage: 1,
  totalPages: 2,
  loadImages: jest.fn(),
  goToNextPage: jest.fn(),
  goToPrevPage: jest.fn(),
};

const images = [{
  url: "",
  width: 20,
  height: 20,
}];

describe("<UnsplashGallery />", () => {
  useUnsplashAPI.mockReturnValue(data);

  it("a click on the search button should toggle the gallery", () => {
    const wrapper = mount(<SearchButton />);
    expect(wrapper.find(".unsplash-gallery__container").length).toBe(0);
    wrapper.find("svg.header__search").simulate("click");
    expect(wrapper.find(".unsplash-gallery__container").length).toBe(1);
    wrapper.find("svg.header__search").simulate("click");
    expect(wrapper.find(".unsplash-gallery__container").length).toBe(0);
  });

  it("a click outside the gallery should close it", () => {
    const map = {};

    document.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });

    const props = {
      close: jest.fn(),
    };

    const menu = mount(<Menu />);
    const wrapper = mount(<UnsplashGallery {... props} />);

    map.click({
      target: ReactDOM.findDOMNode(menu.getDOMNode()),
    });
    expect(props.close).toHaveBeenCalledTimes(1);

    map.click({
      target: ReactDOM.findDOMNode(wrapper.getDOMNode()),
    });
    expect(props.close).toHaveBeenCalledTimes(1);
  });

  it("should display a message when there are no images", () => {
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(wrapper.find(".unsplash-gallery__error")).toHaveLength(1);
    expect(wrapper.find(".unsplash-gallery__error").text()).toBe("No results");

    expect(wrapper.find(".unsplash-gallery__navbar")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__photos")).toHaveLength(0);
  });

  it("when loading images, the loader should be displayed", () => {
    useUnsplashAPI.mockReturnValue({...data, isLoading: true});
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(wrapper.find(".unsplash-gallery__loader")).toHaveLength(1);

    expect(wrapper.find(".unsplash-gallery__navbar")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__photos")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__error")).toHaveLength(0);
  });

  it("should display an error message if something goes wrong", () => {
    useUnsplashAPI.mockReturnValue({...data, isError: true});
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);

    const errorElement = wrapper.find(".unsplash-gallery__error");
    expect(errorElement).toHaveLength(1);
    expect(errorElement.text()).toBe("Something went wrong ...");

    expect(wrapper.find(".unsplash-gallery__loader")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__navbar")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__photos")).toHaveLength(0);
  });

  it("should display images", () => {
    useUnsplashAPI.mockReturnValue({
      ...data,
      images,
      currentPage: 1,
      totalPages: 2,
    });
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(wrapper.find(".unsplash-gallery__photos")).toHaveLength(1);
    expect(wrapper.find(".unsplash-gallery__navbar")).toHaveLength(1);
    expect(wrapper.find(".unsplash-gallery__prev-btn")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__next-btn")).toHaveLength(1);
  });

  it("shouldn't display the navbar if there is only one page", () => {
    useUnsplashAPI.mockReturnValue({
      ...data,
      images,
      currentPage: 1,
      totalPages: 1,
    });
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(wrapper.find(".unsplash-gallery__prev-btn")).toHaveLength(0);
    expect(wrapper.find(".unsplash-gallery__next-btn")).toHaveLength(0);
  });

  it(`both buttons of the navbar should be displayed if the previous 
  and next page exist`, () => {
    useUnsplashAPI.mockReturnValue({
      ...data,
      images,
      currentPage: 2,
      totalPages: 3,
    });
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(wrapper.find(".unsplash-gallery__prev-btn")).toHaveLength(1);
    expect(wrapper.find(".unsplash-gallery__next-btn")).toHaveLength(1);
  });

  it("a click on the navbar buttons should call appropriate functions",
  () => {
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    wrapper.find(".unsplash-gallery__prev-btn").simulate("click");
    expect(data.goToPrevPage).toHaveBeenCalledTimes(1);
    wrapper.find(".unsplash-gallery__next-btn").simulate("click");
    expect(data.goToNextPage).toHaveBeenCalledTimes(1);
  });

  it("shouldn't display the 'next' button if the current page is the last",
  () => {
    useUnsplashAPI.mockReturnValue({
      ...data,
      images,
      currentPage: 3,
      totalPages: 3,
    });
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(wrapper.find(".unsplash-gallery__prev-btn")).toHaveLength(1);
    expect(wrapper.find(".unsplash-gallery__next-btn")).toHaveLength(0);
  });

  it("should pass an input value to the loadImages function", () => {
    const wrapper = shallow(<UnsplashGallery close={jest.fn()}/>);
    wrapper.find(".unsplash-gallery__input").simulate("change", {
      target: {value: "test"},
    });
    expect(wrapper.find(".unsplash-gallery__input").prop("value"))
    .toEqual("test");

    wrapper.find(".unsplash-gallery__form").simulate("submit", {
      preventDefault: jest.fn(),
    });
    expect(data.loadImages).toHaveBeenCalledWith("test");
  });

  it("should match the snapshot", () => {
    useUnsplashAPI.mockReturnValue({...data, images});
    const component = shallow(<UnsplashGallery close={jest.fn()}/>);
    expect(component.html()).toMatchSnapshot();
  });
});