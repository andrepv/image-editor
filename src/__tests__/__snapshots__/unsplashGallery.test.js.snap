// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnsplashGallery /> should match the snapshot 1`] = `"<div class=\\"unsplash-gallery__container\\"><div class=\\"unsplash-gallery custom-scrollbar\\"><form class=\\"unsplash-gallery__form\\"><button type=\\"submit\\" class=\\"unsplash-gallery__submit-btn\\"><svg>search2.svg</svg></button><input type=\\"text\\" class=\\"unsplash-gallery__input\\" value=\\"\\" placeholder=\\"What are you looking for?\\"/></form><div class=\\"unsplash-gallery__navbar\\"><button class=\\"unsplash-gallery__next-btn\\">Next</button></div><div class=\\"unsplash-gallery__photos\\"><div>Â </div></div></div></div>"`;
